cmd/vm-informant/main.go

```diff
@@ -97,54 +102,103 @@ func main() {
 	}
 }
 
+const minWaitDuration = 5 * time.Second
+
 // runRestartOnFailure repeatedly calls this binary with the same flags, but with 'auto-restart'
 // removed.
 //
 // We execute ourselves as a subprocess so that it's possible to appropriately cleanup after
 // termination by various signals (or an unhandled panic!). This is worthwhile because we *really*
 // don't want to leave the cgroup frozen while waiting to restart.
-func runRestartOnFailure(args []string, cleanupHooks []func()) {
+func runRestartOnFailure(ctx context.Context, args []string, cleanupHooks []func()) {
 	selfPath := os.Args[0]
-
-	minWaitDuration := time.Second * 5
+	timer := time.NewTimer(0)
+	defer timer.Stop()
 
 	for {
 		startTime := time.Now()
-
-		cmd := exec.CommandContext(context.TODO(), selfPath, args...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		err := cmd.Run()
-
+		sig := make(chan struct{})
 		var exitMode string
 
-		if err != nil {
-			// lint note: the linter's worried about wrapped errors being incorrect with switch, but
-			// this is cleaner than the alternative (via errors.As) and it's still correct because
-			// exec.Command.Run() explicitly mentions ExitError.
-			switch err.(type) { //nolint:errorlint // see above.
-			case *exec.ExitError:
-				exitMode = "failed"
-				klog.Errorf("vm-informant exited with: %v", err)
-			default:
-				exitMode = "failed to start"
-				klog.Errorf("error running vm-informant: %v", err)
+		func() {
+			pctx, pcancel := context.WithCancel(context.Background())
+			defer pcancel()
+
+			cmd := exec.Command(selfPath, args...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			go func() {
+				defer close(sig)
+				select {
+				case <-pctx.Done():
+					return
+				case <-ctx.Done():
+					if cmd.Process == nil || cmd.ProcessState == nil {
+						return
+					}
```

Comment:
This is racy. Aside from data races around `cmd.Process` and `cmd.ProcessState` (which shouldn't have a material impact beyond annoying the race detector), it's possible for the following sequence of events to occur:

1. Spawn this function
2. Parent vm-informant receives SIGTERM, this function sees `cmd.Process == nil` and returns
3. `cmd.Run()` starts
4. Child vm-informant never receives any signal.

We could partly fix this by putting `defer pcancel()` inside this goroutine. However:

AFAICT any solution using `cmd.Run()` will be similarly racy. I'd rewrite with `Start` + `Wait` instead.

Reply:
Yeah, I think this is a big part of why I didn't want us to have to be in the business of doing this kind of process management: there are a lot of corners. I don't know that Start+Wait definitely gets you out of the potential of 4 happening, (because what I think you really want is to run the subprocess in a cgroup that will terminate on the parent process exiting.) This us (unfortunately) a thing that windows has done better than Linux (with job objects), and you might be able to get it with cgroups/systemd managed processes, but it's dodgy to do this "correctly" without help from the operating system.